File: Data Structures & Algorithms/top-k-elements-in-list/submission-11.cpp
#include <algorithm>
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        std::unordered_map<int, int> counter;
        std::vector<int> results;

        for (const int &num: nums) {
            counter[num]++;
        }

        std::vector<vector<int>> frequencies (nums.size());

        for (auto pair: counter) {
            frequencies[pair.second - 1].push_back(pair.first);
        }

        
        for ( int i = nums.size() - 1; i >= 0; --i ) {
            for (int num: frequencies[i]) {
                if (k == 0) return results;
                else {
                    results.push_back(num);
                    k--;
                }
            }
        }
        
        return results;
    }
};

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-11.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:

        results = [1] * len(nums)
        prefix = nums[0]

        for i in range(1, len(nums)):
            results[i] = prefix
            prefix *= nums[i]

        suffix = nums[-1]

        for i in range(len(nums) - 2, -1, -1):
            results[i] *= suffix
            suffix *= nums[i]
        
        return results
        
        